authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,134 @@
 #  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  * See the License for the specific language governing permissions and
 #  * limitations under the License.
-
+import logging
 
 import ldap
 from flask_securest.authentication_providers.abstract_authentication_provider \
     import AbstractAuthenticationProvider
 from flask_securest import utils
+from flask_securest import constants
+from flask_securest.exceptions import AuthenticationException
+
+DN_ATTRIBUTE_NAME = 'DN'
+DEFAULT_AUTH_ATTR_NAME = 'uid'
 
 
 class LDAPAuthenticationProvider(AbstractAuthenticationProvider):
-    def __init__(self, directory_url):
-        self.directory_url = directory_url
+    def __init__(self, ldap_url, search_props=None):
+        self.lgr = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+        self.ldap_url = ldap_url
+        if search_props:
+            self.base_dn = search_props['base_dn']
+            self.admin_user_id = search_props['admin_user_id']
+            self.admin_password = search_props['admin_password']
+            self.user_id_attribute = search_props.get('user_id_attribute',
+                                                      DEFAULT_AUTH_ATTR_NAME)
 
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_search_based_authentication():
+            self.authenticate_by_attribute(who, password)
+        else:
+            self.validate_bind(who, password)
+        return who
+
+    def authenticate_by_attribute(self, who, password):
+        try:
+            conn = self.create_bind_conn(self.admin_user_id,
+                                         self.admin_password)
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_dn: {1}, provided in the configuration; {2}' \
+                .format(self.ldap_url, who, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
+
+        attribute_name = self.user_id_attribute
```

Comment:
Since this is a utility method it does not have to be aware of properties set on self. also you can reduce the 2-lines to one: search_filter = '{0}={1}'.format(identifying_attribute, value)